README.md
# Haskell Chat (IRC) App
This is a haskell stack application

### To Run Server
Use the stack command "stack run server"

### To Run Client
You must first update the src/Client.hs file and add your computers IP Address

This line specifically: connect sock $ SockAddrInet 3000 $ tupleToHostAddress (xxx, xxx, xxx, xxx)

Use the stack command "stack run client"

## Example Client Commands
/whisper Alice Hey Alice -- Allows private messaging between two clients

/join BasketBall Hockey ComputerScience -- Allows the client to join specified chat rooms

/leave Hockey ComputerScience -- Allows client to leave specified chat rooms

/name Bob -- Allows client to re-name themselves

/myName -- Gets your saved name from the server

/chatRooms -- Allows client to list all active chat rooms on server

/myChatRooms -- Allows client to list thier own chat rooms

/sendTo BasketBall Did you see the game last night? -- Allows messages to a single specified chat room

/members BasketBall -- Allows client to get list of all members of a chat room

/quit -- Allows client to disconnect from server

/test -- Runs client side test suite

## Example Server Commands
/sendTo BasketBall The Blazers are up by 10! -- Allows the server to send messages to a specified chat room

/sendToAll Shutting down in 15 minutes. -- Allows the server to send messages to all clients

/disconnect Bob -- Allows the server to disconnect specific clients

/disconnectAll -- Allows the server to disconnect all clients

/quit -- Shuts down server.


CHANGELOG.md
# Changelog for `newApp`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Client.hs
module Client 
    ( mainClient
    ) where

import Network.Socket
import Network.Socket.ByteString
import qualified Data.ByteString.Char8 as Byte
import System.IO
import Control.Concurrent
import Data.List
import Control.Exception

mainClient :: IO ()
mainClient = withSocketsDo $ do
  -- create a TCP socket
  sock <- socket AF_INET Stream defaultProtocol

  -- connect to the server

  -- Desktop IP Address
  --connected <- try (connect sock $ SockAddrInet 3000 $ tupleToHostAddress (192, 168, 1, 8)) :: IO (Either SomeException ())

  -- Laptop IP Address @ School
  connected <- try (connect sock $ SockAddrInet 3000 $ tupleToHostAddress (10, 200, 224, 219)) :: IO (Either SomeException ())
  case connected of
    Left e -> putStrLn $ "Error connecting: " ++ show e
    Right _ -> do 
      putStrLn "Connected to server!"
      loop sock


  -- read and sendAll messages to the server until the user types "/quit"

loop :: Socket -> IO ()
loop sock = do
  msgVar <- newMVar []
  -- Get name from client
  putStrLn "Please enter your name."
  name <- getLine

  trySend <- try (sendAll sock (Byte.pack ("/init " ++ name))) :: IO (Either SomeException ())
  case trySend of
    Left e -> putStrLn $ "Error initiating name: " ++ show e
    Right _ -> pure ()

  sendThread <- forkIO $ sendLoop msgVar
  receiveLoop sendThread msgVar

  -- sendLoop needs to be non-blocking IO so when server closes connection the client doen't hang
  where
    sendLoop msgVar = do
      inputReady <- hWaitForInput stdin 0
      if inputReady
        then do
          msg <- hGetLine stdin
          case msg of
            -- sending signal to close connection from client side
            "/quit" -> do
             sendAll sock $ Byte.pack "exit"
             putStrLn "Goodbye!"
            -- Start the test suite
            "/test" -> do
              testJoinRooms sock msgVar
              testLeaveChatRooms sock msgVar
              testChangeName sock msgVar
              sendLoop msgVar
            -- Otherwise send message and loop
            _ -> do
              trySend' <- try (sendAll sock (Byte.pack msg)) :: IO (Either SomeException ())
              case trySend' of
                Left e -> putStrLn $ "Error sending message: " ++ show e
                Right _ -> pure ()
              sendLoop msgVar
        else do sendLoop msgVar

    receiveLoop sendThread msgVar = do
      response <- try (recv sock 1024) :: IO (Either SomeException Byte.ByteString)
      case response of
        Left e -> putStrLn $ "Server Connection Closed: " ++ show e
        Right m -> do
          let msg = Byte.unpack m
          if Byte.null m
            then putStrLn "Server closed the connection."
            else do
              case msg of
                -- Server force closing connection
                "exit" -> do
                  sendAll sock $ Byte.pack "exit"
                  putStrLn "Server closed the connection."
                  killThread sendThread
                -- Otherwise print message and loop
                _ -> do
                  putStrLn msg
                  modifyMVar_ msgVar (\messages -> return (msg : messages))
                  receiveLoop sendThread msgVar

-- Test Functions
testJoinRooms :: Socket -> MVar [String] -> IO ()
testJoinRooms sock msgVar = do
  sendAll sock $ Byte.pack "/myChatRooms"
  -- Need to wait for server response
  threadDelay 1000000
  messages <- readMVar msgVar
  let originalRooms = read (head messages) :: [String]
  sendAll sock $ Byte.pack "/join"
  -- Need to wait for server response
  threadDelay 1000000
  sendAll sock $ Byte.pack "/myChatRooms"
  -- Need to wait for server response
  threadDelay 1000000
  messages' <- readMVar msgVar
  let newRooms = read (head messages') :: [String]
  if newRooms == originalRooms
    then do
      let rooms = "Room1 Room2 1502 BasketBall ComputerScience Testing"
      let wordRooms = words rooms
      sendAll sock $ Byte.pack $ "/join " ++ rooms
      -- Need to wait for server response
      threadDelay 1000000
      sendAll sock $ Byte.pack "/myChatRooms"
      -- Need to wait for server response
      threadDelay 1000000
      messages'' <- readMVar msgVar
      let newRooms' = read (head messages'') :: [String]
      if newRooms' == (wordRooms ++ originalRooms)
        then putStrLn "Test Join: Passed"
      else putStrLn "Test Join: Failed"
    else putStrLn "Test Join: Failed"

testLeaveChatRooms :: Socket -> MVar [String] -> IO ()
testLeaveChatRooms sock msgVar = do
  let command = "/myChatRooms"
  sendAll sock $ Byte.pack command
  -- Wait for server response
  threadDelay 1000000
  messages <- readMVar msgVar
  let originalRooms = read (head messages) :: [String]
  sendAll sock $ Byte.pack "/leave"
  -- Wait for server response
  threadDelay 1000000
  sendAll sock $ Byte.pack command
  -- Wait for server response
  threadDelay 1000000
  messages' <- readMVar msgVar
  let newRooms = read (head messages') :: [String]
  if newRooms == originalRooms
    then do
      let rooms = "Room2 1502 BasketBall Testing"
      let wordRooms = words rooms
      sendAll sock $ Byte.pack $ "/leave " ++ rooms
      -- Wait for server response
      threadDelay 1000000
      sendAll sock $ Byte.pack command
      -- Wait for server response
      threadDelay 1000000
      messages'' <- readMVar msgVar
      let newRooms' = read (head messages'') :: [String]
      if newRooms' == (originalRooms \\ wordRooms)
        then putStrLn "Test Leave: Passed"
      else putStrLn "Test Leave: Failed"
    else putStrLn "Test Leave: Failed"

testChangeName :: Socket -> MVar [String] -> IO ()
testChangeName sock msgVar = do
  let name = "This is a test"
  sendAll sock $ Byte.pack $ "/name " ++ name
  -- Wait for server response
  threadDelay 1000000
  sendAll sock $ Byte.pack "/myName"
  -- Wait for server response
  threadDelay 1000000
  messages <- readMVar msgVar
  let msg = drop (length (head messages) - length name) (head messages)
  if msg == name
    then do
      let name' = "This is also a test. Testing... Testing..."
      sendAll sock $ Byte.pack $ "/name " ++ name'
      -- Wait for server response
      threadDelay 1000000
      sendAll sock $ Byte.pack "/myName"
      -- Wait for server response
      threadDelay 1000000
      messages' <- readMVar msgVar
      let msg' = drop (length (head messages') - length name') (head messages')
      if msg' == name'
        then do 
          let name'' = ""
          sendAll sock $ Byte.pack $ "/name " ++ name''
          -- Wait for server response
          threadDelay 1000000
          messages'' <- readMVar msgVar
          let msg'' = head messages''
          if msg'' == "Name was empty, save unsuccessfull."
            then putStrLn "Test Name: Passed"
          else putStrLn "Test Name: Failed"
      else putStrLn "Test Name: Failed"
  else putStrLn "Test Name: Failed"


src/Server.hs
module Server 
    ( mainServer
    ) where

{-# LANGUAGE OverloadedStrings #-}

import Network.Socket
import Network.Socket.ByteString
import qualified Data.ByteString.Char8 as Byte
import Data.Maybe
import Control.Monad
import Control.Concurrent 
import Data.List
import Control.Exception
import System.IO

data ClientInfo = ClientInfo 
  {
    clientSocket :: Socket,
    clientName :: String,
    clientRooms :: [String]
  } deriving (Eq)

mainServer :: IO ()
mainServer = withSocketsDo $ do
  serverSock <- socket AF_INET Stream 0
  setSocketOption serverSock ReuseAddr 1
  bind serverSock (SockAddrInet 3000 0)
  listen serverSock 2

  putStrLn "Listening on port 3000"

  clientInfo <- newMVar []

  -- Fork acceptLoop so server can listen for commands
  _ <- forkIO $ acceptLoop serverSock clientInfo

  mainLoop serverSock clientInfo

-- Forked loop that listens for server commands, while the server goes on to accept clients
mainLoop :: Socket -> MVar [ClientInfo] -> IO ()
mainLoop sock clientInfo = do
  inputReady <- hWaitForInput stdin 0
  if inputReady
    then do
      msg <- hGetLine stdin
      let wordsMsg = words msg
      if not (null wordsMsg)
        then do
          case head wordsMsg of
            -- Closes server
            "/quit" -> do
              putStrLn "Closing Server"
              close sock
            -- Allows server communication to specified chat room
            "/sendTo" -> do
              let room = wordsMsg !! 1
              clientList <- readMVar clientInfo
              let msgToSend = "(Server): " ++ drop (length (head wordsMsg) + length (wordsMsg !! 1) + 2) msg
              mapM_ (\c -> when (room `elem` clientRooms c) $ broadcast (clientSocket c) msgToSend) clientList
              mainLoop sock clientInfo
            -- Allows server to communicate to all clients
            "/sendToAll" -> do
              clientList <- readMVar clientInfo
              let msgToSend = "(Server): " ++ drop (length (head wordsMsg) + 1) msg
              mapM_ (\c -> broadcast (clientSocket c) msgToSend) clientList
              mainLoop sock clientInfo
            -- Disconnects a specific client from the server
            "/disconnect" -> do
              let client = drop (length (head wordsMsg) + 1) msg
              clientList <- readMVar clientInfo
              let clientMaybe = getClientByName client clientList
              if isJust clientMaybe
                then broadcast (clientSocket (getMaybeClient clientMaybe)) "exit"
              else putStrLn "Client not found."
              mainLoop sock clientInfo
            "/disconnectAll" -> do
              clientList <- readMVar clientInfo
              mapM_ (\c -> broadcast (clientSocket c) "exit") clientList
              mainLoop sock clientInfo
            _ -> do
              mainLoop sock clientInfo
        else do mainLoop sock clientInfo
    else do mainLoop sock clientInfo

-- Accept client loop, adds clients to the client MVar
acceptLoop :: Socket -> MVar [ClientInfo] -> IO ()
acceptLoop serverSock clientInfo = do
  (sock, addr) <- accept serverSock
  putStrLn $ "Accepted connection from " ++ show addr

  -- add the new client to the list
  modifyMVar_ clientInfo $ \clientInfos -> return ( ClientInfo sock "" ["General"] : clientInfos)

  -- fork a new thread to handle the client
  _ <- forkIO $ handleClient sock clientInfo

  -- continue accepting new clients
  acceptLoop serverSock clientInfo

-- The forked individual client handler
handleClient :: Socket -> MVar [ClientInfo] -> IO ()
handleClient sock clientInfo = do
  -- loop to receive and broadcast messages
  loop
  where
    loop = do
      clientList' <- readMVar clientInfo
      let currentClient = getCurrentClient sock clientList'
      msg <- try (recv sock 1024) :: IO (Either SomeException Byte.ByteString)
      case msg of
        Left e -> do
          putStrLn $ "Client " ++ clientName currentClient ++ " Disconneted: " ++ show e
          modifyMVar_ clientInfo $ \clients -> return (removeClients currentClient clients)
          close sock
        Right m -> do
          let msg' = Byte.unpack m
          let wordsMsg = words msg'
          case head wordsMsg of
            -- Save inital name givin to server
            "/init" -> do
              let name = drop (length (head wordsMsg) + 1) msg'
              putStrLn (name ++ " has entered the server!")
              modifyMVar_ clientInfo $ \clients -> return (editClientName (ClientInfo sock name (clientRooms currentClient)) clients)
              broadcast (clientSocket currentClient) "Successfully saved name."
              threadDelay 250000
              broadcast (clientSocket currentClient) "Joined chat room [\"General\"]."
              loop
            -- Private message between two clients
            "/whisper" -> do
              clientList <- readMVar clientInfo
              let maybeClient = getClientByMsg msg' clientList
              if isJust maybeClient 
                then do
                  let client = getMaybeClient maybeClient
                  let msg'' = drop (length (head wordsMsg) + length (clientName client) + 2) msg'
                  broadcast (clientSocket client) ("(Whisper) " ++ clientName currentClient ++ ": " ++ msg'')
              else broadcast (clientSocket currentClient) "Client not found."
              loop
            -- Lets client re-name themselves
            "/name" -> do
              let name = drop (length (head wordsMsg) + 1) msg'
              if not (null name)
                then do
                  putStrLn (clientName currentClient ++ " has changed thier name to " ++ name)
                  modifyMVar_ clientInfo $ \clients -> return (editClientName (ClientInfo sock name (clientRooms currentClient)) clients)
                  broadcast (clientSocket currentClient) "Successfully saved name."
                else broadcast (clientSocket currentClient) "Name was empty, save unseccessfull."
              loop
            -- Lets client check current saved name
            "/myName" -> do
              let name = clientName currentClient
              broadcast (clientSocket currentClient) ("Your saved name is: " ++ name)
              loop
            -- List all current client chat rooms
            "/myChatRooms" -> do
              broadcast (clientSocket currentClient) (show (clientRooms currentClient))
              loop
            -- Lists all active chat rooms on server
            "/chatRooms" -> do
              clientList <- readMVar clientInfo
              broadcast (clientSocket currentClient) (show (listChatRooms clientList []))
              loop
            -- Allows client to sent to a specific chat room
            "/sendTo" -> do
              clientList <- readMVar clientInfo
              let room = wordsMsg !! 1
              let msgToSend = "[" ++ show room ++ "] " ++ clientName currentClient ++ ": " ++ drop (length (head wordsMsg) + length (wordsMsg !! 1) + 2) msg'
              mapM_ (\c -> when (clientSocket c /= sock && room `elem` clientRooms c) $ broadcast (clientSocket c) msgToSend ) clientList
              loop
            -- Lists all members of specific chat room
            "/members" -> do
              let room = wordsMsg !! 1
              clientList <- readMVar clientInfo
              broadcast (clientSocket currentClient) (show (listClients room clientList))
              loop
            -- Removes client from desired chat room/s
            "/leave" -> do
              let rooms = drop 1 wordsMsg
              if rooms /= []
                then do
                  clientList <- readMVar clientInfo
                  mapM_ (\c -> when (clientSocket c /= sock && hasCommonElements (clientRooms c) rooms && clientRooms c /= []) 
                    $ broadcast (clientSocket c) (clientName currentClient ++ " has left the chat room!")) clientList
                  modifyMVar_ clientInfo $ \clients -> return (leaveClientRooms (ClientInfo sock "" rooms) clients)
                  broadcast (clientSocket currentClient) ("Successfully left chat rooms: " ++ show rooms)
                else broadcast (clientSocket currentClient) ("Please enter a chat room to leave.")
              loop
            -- Adds client to desired chat room/s
            "/join" -> do
              let rooms = drop 1 wordsMsg
              if rooms /= []
                then do
                  modifyMVar_ clientInfo $ \clients -> return (addClientRooms (ClientInfo sock "" rooms) clients)
                  clientList <- readMVar clientInfo
                  mapM_ (\c -> when (clientSocket c /= sock && hasCommonElements (clientRooms c) rooms && clientRooms c /= []) 
                    $ broadcast (clientSocket c) (clientName currentClient ++ " has joined the chat room!")) clientList
                  broadcast (clientSocket currentClient) ("Successfully joined chat rooms: " ++ show rooms)
                else broadcast (clientSocket currentClient) ("Please enter a chat room to join.")
              loop
            -- Client terminated connection
            "exit" -> do
              putStrLn  $ clientName currentClient ++ " has disconnected from server."
              modifyMVar_ clientInfo $ \clients -> return (removeClients currentClient clients)
              close sock
            -- Relay message. Don't send to self, or chat rooms you aren't in. Clients in empty chat rooms shouldn't see
            -- into other chat rooms, but should see other clients in empty chat rooms
            _ -> do
              clientList <- readMVar clientInfo
              putStrLn (clientName currentClient ++ ": " ++ msg')
              mapM_ (\c -> when (clientSocket c /= sock && hasCommonElements (clientRooms c) (clientRooms currentClient)
                && (null (clientRooms c) && null (clientRooms currentClient) || clientRooms c /= [] && clientRooms currentClient /= [])) 
                $ broadcast (clientSocket c) (show (getMatchingRooms (clientRooms c) (clientRooms currentClient)) 
                ++ " " ++ clientName currentClient ++ ": " ++ msg')) clientList
              loop

-- Helper functions

-- broadcast using exception handling
broadcast :: Socket -> String -> IO ()
broadcast sock msg = do
  trySend <- try (sendAll sock (Byte.pack msg)) :: IO (Either SomeException ())
  case trySend of
    Left e -> putStrLn $ "Error sending message: " ++ show e
    Right _ -> pure ()

-- Check if two lists of chat rooms have common elements
hasCommonElements :: Eq a => [a] -> [a] -> Bool
hasCommonElements xs ys = not (null (intersect xs ys))

-- Gets all active chat rooms
listChatRooms :: [ClientInfo] -> [String] -> [String]
listChatRooms [] [] = []
listChatRooms [] rooms = nub rooms
listChatRooms (c:cs) rooms = listChatRooms cs (clientRooms c ++ rooms)

-- Lists all clients in a chatroom
listClients :: String -> [ClientInfo] -> [String]
listClients room clientList = do
  currentClient <- clientList
  if room `elem` clientRooms currentClient
    then return $ clientName currentClient
  else []

-- Gets matching chat rooms from two clients rooms
getMatchingRooms :: [String] -> [String] -> [String]
getMatchingRooms xs ys = filter (`elem` ys) xs

-- Removes client from list of clients
removeClients :: ClientInfo -> [ClientInfo] -> [ClientInfo]
removeClients toRemove clientList = do
  currentClient <- clientList
  if clientSocket currentClient /= clientSocket toRemove 
    then return currentClient
  else []

-- Updates a clients name
editClientName :: ClientInfo -> [ClientInfo] -> [ClientInfo]
editClientName toEdit clientList = do
  currentClient <- clientList
  if clientSocket currentClient == clientSocket toEdit
    then return toEdit
  else return currentClient

--  Updates a clients chat rooms
addClientRooms :: ClientInfo -> [ClientInfo] -> [ClientInfo]
addClientRooms toAdd clientList = do
  currentClient <- clientList
  if clientSocket currentClient == clientSocket toAdd
    then return $ ClientInfo (clientSocket currentClient) (clientName currentClient) 
      (addClientRoomsHelper (clientRooms toAdd) (clientRooms currentClient))
  else return currentClient

addClientRoomsHelper :: [String] -> [String] -> [String]
addClientRoomsHelper [] [] = []
addClientRoomsHelper [] xs = xs
addClientRoomsHelper xs [] = xs
addClientRoomsHelper (x:xs) (y:ys) | x `elem` (y:ys) = addClientRoomsHelper xs (y:ys)
                                   | otherwise = x : addClientRoomsHelper xs (y:ys)

-- Gets current client when in client handler
getCurrentClient :: Socket -> [ClientInfo] -> ClientInfo
getCurrentClient sock [] = ClientInfo sock "" []
getCurrentClient sock (c:cs) | clientSocket c == sock = c
                             | otherwise = getCurrentClient sock cs

-- Matches a client name from a whole message string (For whisper)
getClientByMsg :: String -> [ClientInfo] -> Maybe ClientInfo
getClientByMsg _ [] = Nothing
getClientByMsg [] _ = Nothing
getClientByMsg msg (c:cs) | clientName c `isInfixOf` msg = Just c
                          | otherwise = getClientByMsg msg cs

-- Gets a client by name
getClientByName :: String -> [ClientInfo] -> Maybe ClientInfo
getClientByName _ [] = Nothing
getClientByName [] _ = Nothing
getClientByName name (c:cs) | clientName c == name = Just c
                            | otherwise = getClientByName name cs

-- Non exaustive, always check if the maybe is not nothing before using.
getMaybeClient :: Maybe ClientInfo -> ClientInfo
getMaybeClient (Just client) = client

-- Updates client rooms
leaveClientRooms :: ClientInfo -> [ClientInfo] -> [ClientInfo]
leaveClientRooms toLeave clientList = do
  currentClient <- clientList
  if clientSocket currentClient == clientSocket toLeave
    then return $ ClientInfo (clientSocket currentClient) (clientName currentClient) 
      (leaveClientRoomsHelper (clientRooms toLeave) (clientRooms currentClient))
  else return currentClient

leaveClientRoomsHelper :: [String] -> [String] -> [String]
leaveClientRoomsHelper [] [] = []
leaveClientRoomsHelper _ [] = []
leaveClientRoomsHelper [] ys = ys
leaveClientRoomsHelper xs (y:ys) | y `elem` xs = leaveClientRoomsHelper (delete y xs) ys
                                 | otherwise = y : leaveClientRoomsHelper xs ys
